Connecting to NTRIP fails on exception UnicodeDecodeError

**Describe the bug**

GNSSNTRIPClient does not work if the server does not return the header in utf-8, but e.g. in cp1250, such as czeposr.cuzk.gov.cz. Connecting to NTRIP fails on exception UnicodeDecodeError('utf-8', b'ICY 200 OK\r\nServer: GNSS Spider 7.9.0.386/1.0\r\nDate: \xfat, 08 \xf8\xedj 2024 13:04:43 GMT', 53, 54, 'invalid start byte')

`2024-10-08 16:35:00,817 [ERROR] pygnssutils.gnssntripclient: ConnectionAbortedError("Unable to parse response headers - 'utf-8' codec can't decode byte 0xfa in position 53: invalid start byte"). Retrying in 10 secs (1/5) ...`


https://github.com/semuconsulting/pygnssutils/blob/85bbc16fe8ccdcbe5c91ba34ff3e57719c6ead0b/src/pygnssutils/gnssntripclient.py#L543

The bug has been showing up since version 1.1.0

**To Reproduce**

Connect to an NTRIP server that returns unsupported utf-8 (or ASCII) characters in the header.

**Expected Behaviour**

A quick fix might be to replace it with this one
`hdr = hdr.decode(errors='backslashreplace').split("\r\n")`

--- a/.github/workflows/checkpr.yml
+++ b/.github/workflows/checkpr.yml
@@ -8,7 +8,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.9, "3.10", "3.11", "3.12", "3.13.0-rc.3"]
+        python-version: [3.9, "3.10", "3.11", "3.12", "3.13"]
 
     steps:
       - uses: actions/checkout@v4

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -10,7 +10,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.9, "3.10", "3.11", "3.12", "3.13.0-rc.3"]
+        python-version: [3.9, "3.10", "3.11", "3.12", "3.13"]
 
     steps:
       - uses: actions/checkout@v4

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -2,9 +2,8 @@
     "python.testing.pytestEnabled": true,
     "python.testing.unittestEnabled": false,
     "editor.formatOnSave": true,
-    "python3.9InterpreterPath": "/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9",
+    "python.defaultInterpreterPath": "python3",
     "modulename": "${workspaceFolderBasename}",
     "distname": "${workspaceFolderBasename}",
-    "moduleversion": "1.1.4",
-    "python.defaultInterpreterPath": "python3"
+    "moduleversion": "1.1.4"
 }

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -5,5 +5,5 @@
     "python3.8InterpreterPath": "/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8",
     "modulename": "${workspaceFolderBasename}",
     "distname": "${workspaceFolderBasename}",
-    "moduleversion": "1.1.3"
+    "moduleversion": "1.1.4"
 }

--- a/.vscode/tasks.json
+++ b/.vscode/tasks.json
@@ -124,16 +124,6 @@
             "problemMatcher": []
         },
         {
-            "label": "Test3.8",
-            "type": "process",
-            "command": "${config:python3.8InterpreterPath}",
-            "args": [
-                "-m",
-                "pytest",
-            ],
-            "problemMatcher": []
-        },
-        {
             "label": "Sphinx",
             "type": "process",
             "command": "sphinx-apidoc",

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -1,5 +1,11 @@
 # pygnssutils Release Notes
 
+### RELEASE 1.1.4
+
+ENHANCEMENTS:
+
+1. Minor enhancements to ubxsetrate CLI utility.
+
 ### RELEASE 1.1.3
 
 ENHANCEMENTS:

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -2,9 +2,9 @@
 
 ### RELEASE 1.1.4
 
-ENHANCEMENTS:
+FIXES:
 
-1. Minor enhancements to ubxsetrate CLI utility.
+1. Fix UnicodeDecodeError in gnssntripclient - Fixes [#93](https://github.com/semuconsulting/pygnssutils/issues/93).
 
 ### RELEASE 1.1.3
 

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -5,6 +5,7 @@
 FIXES:
 
 1. Fix UnicodeDecodeError in gnssntripclient - Fixes [#93](https://github.com/semuconsulting/pygnssutils/issues/93).
+1. Fix occasional ubxload 'lockups'.
 
 ### RELEASE 1.1.3
 

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -2,10 +2,11 @@
 
 ### RELEASE 1.1.4
 
-FIXES:
+ENHANCEMENTS:
 
 1. Fix UnicodeDecodeError in gnssntripclient - Fixes [#93](https://github.com/semuconsulting/pygnssutils/issues/93).
-1. Fix occasional ubxload 'lockups'.
+1. Fix occasional `ubxload` 'lockups' - Fixes [#48](https://github.com/semuconsulting/pygnssutils/issues/48).
+1. Update `ubx...` utilities to use common logging provisions.
 
 ### RELEASE 1.1.3
 

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -7,6 +7,8 @@ ENHANCEMENTS:
 1. Fix UnicodeDecodeError in gnssntripclient - Fixes [#93](https://github.com/semuconsulting/pygnssutils/issues/93).
 1. Fix occasional `ubxload` 'lockups' - Fixes [#48](https://github.com/semuconsulting/pygnssutils/issues/48).
 1. Update `ubx...` utilities to use common logging provisions.
+1. Drop active support for Python 3.8 - End of Life as at October 2024.
+1. Update min pyubx2 version to 1.2.47.
 
 ### RELEASE 1.1.3
 

--- a/docs/conf.py
+++ b/docs/conf.py
@@ -58,7 +58,20 @@ html_title = "<project> v<version> documentation."
 # html_static_path = ["_static"]
 html_last_updated_fmt = "%b %d %Y"
 html_theme_options = {
-    "display_version": True,
+    "logo_only": False,
+    "prev_next_buttons_location": "bottom",
+    "style_external_links": False,
+    "vcs_pageview_mode": "",
+    "style_nav_header_background": "white",
+    "flyout_display": "hidden",
+    "version_selector": True,
+    "language_selector": True,
+    # Toc options
+    "collapse_navigation": True,
+    "sticky_navigation": True,
+    "navigation_depth": 4,
+    "includehidden": True,
+    "titles_only": False,
 }
 
 autodoc_default_options = {

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -7,7 +7,7 @@ name = "pygnssutils"
 authors = [{ name = "semuadmin", email = "semuadmin@semuconsulting.com" }]
 maintainers = [{ name = "semuadmin", email = "semuadmin@semuconsulting.com" }]
 description = "GNSS Command Line Utilities"
-version = "1.1.3"
+version = "1.1.4"
 license = { file = "LICENSE" }
 readme = "README.md"
 requires-python = ">=3.8"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,7 +37,7 @@ dependencies = [
     "certifi>=2024.0.0",
     "paho-mqtt>=1.6.0",  # waiting for >=2.0.0 in Conda
     "pyserial>=3.5",
-    "pyspartn>=1.0.3",
+    "pyspartn>=1.0.4",
     "pyubx2>=1.2.47",
 ]
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -38,7 +38,7 @@ dependencies = [
     "paho-mqtt>=1.6.0",  # waiting for >=2.0.0 in Conda
     "pyserial>=3.5",
     "pyspartn>=1.0.3",
-    "pyubx2>=1.2.46",
+    "pyubx2>=1.2.47",
 ]
 
 [project.scripts]

--- a/src/pygnssutils/_version.py
+++ b/src/pygnssutils/_version.py
@@ -8,4 +8,4 @@ Created on 2 Oct 2020
 :license: BSD 3-Clause
 """
 
-__version__ = "1.1.3"
+__version__ = "1.1.4"

--- a/src/pygnssutils/globals.py
+++ b/src/pygnssutils/globals.py
@@ -178,7 +178,7 @@ MINUBX = "minubx"
 ALLNMEA_CLS = [b"\xF0", b"\xF1"]
 MINMMEA_ID = [b"\xF0\x00", b"\xF0\x02", b"\xF0\x03", b"\xF0\x04", b"\xF0\x05"]
 ALLUBX_CLS = [b"\x01"]
-MINUBX_ID = [b"\x01\x07", b"\x01\x35"]
+MINUBX_ID = [b"\x01\x04", b"\x01\x07", b"\x01\x35"]
 
 TOPIC_KEY = "/pp/ubx/0236/{}"
 TOPIC_ASSIST = "/pp/ubx/mga"

--- a/src/pygnssutils/gnssntripclient.py
+++ b/src/pygnssutils/gnssntripclient.py
@@ -475,7 +475,8 @@ class GNSSNTRIPClient:
                     hdr, bdy = hdrbdy[0], b""
             else:  # has body content
                 hdr, bdy = hdrbdy
-            hdr = hdr.decode().split("\r\n")
+            # some legacy casters use cp1250 rather than utf-8
+            hdr = hdr.decode(errors="backslashreplace").split("\r\n")
             status = hdr[0].split(" ", 3)
             self._response_status = {
                 "protocol": status[0],

--- a/src/pygnssutils/gnssntripclient_cli.py
+++ b/src/pygnssutils/gnssntripclient_cli.py
@@ -141,7 +141,7 @@ def main():
         "--datatype",
         required=False,
         help="Data type (RTCM or SPARTN)",
-        choices=[RTCM, "rtcm", SPARTN, "spartn"],
+        choices=[RTCM, RTCM.upper(), SPARTN, SPARTN.upper()],
         default=RTCM,
     )
     ap.add_argument(
@@ -186,7 +186,7 @@ def main():
     ap.add_argument(
         "--spartnkey",
         required=False,
-        help="Decryption key for encrypted SPARTN payloads",
+        help="Hexadecimal decryption key for encrypted SPARTN payloads",
         default=getenv(ENV_MQTT_KEY, default=None),
     )
     ap.add_argument(

--- a/src/pygnssutils/gnssstreamer.py
+++ b/src/pygnssutils/gnssstreamer.py
@@ -17,7 +17,7 @@ from io import UnsupportedOperation
 from logging import getLogger
 from queue import Empty, Queue
 from threading import Event, Thread
-from time import sleep, time
+from time import time
 
 from pynmeagps import NMEAMessage, NMEAParseError
 from pyrtcm import RTCMMessage, RTCMParseError

--- a/src/pygnssutils/ubxcompare.py
+++ b/src/pygnssutils/ubxcompare.py
@@ -21,6 +21,7 @@ Outputs dictionary of config keys and their values for each file e.g.
 """
 
 from argparse import ArgumentDefaultsHelpFormatter, ArgumentParser
+from logging import getLogger
 
 from pyubx2 import (
     POLL_LAYER_BBR,
@@ -38,7 +39,8 @@ from pyubx2 import (
 )
 
 from pygnssutils._version import __version__ as VERSION
-from pygnssutils.globals import EPILOG
+from pygnssutils.globals import EPILOG, VERBOSITY_HIGH
+from pygnssutils.helpers import set_common_args
 
 CFG = b"\x06"
 VALGET = b"\x8b"
@@ -59,6 +61,7 @@ class UBXCompare:
         :param bool diffsonly: True = show diffs only, False = show entire config (True)
         """
 
+        self.logger = getLogger(__name__)
         if infiles in ("", None):
             raise ValueError("--infiles parameter must not be blank")
 
@@ -72,9 +75,9 @@ class UBXCompare:
             fcount += 1
             self.parse_file(cfgdict, file.strip(), fcount, form)
 
-        print(
-            f"\n{fcount} files processed, list of {'differences in' if diffsonly else 'all'}",
-            "config keys and their values follows:\n",
+        self.logger.info(
+            f"{fcount} files processed, list of {'differences in' if diffsonly else 'all'}"
+            " config keys and their values follows: ",
         )
 
         for key, vals in dict(sorted(cfgdict.items())).items():
@@ -87,7 +90,7 @@ class UBXCompare:
             if (diffsonly and diff) or not diffsonly:
                 print(f"{key} ({'DIFFS!' if diff else None}); {str(vals).strip('{}')}")
 
-        print(f"\nTotal config keys: {kcount}. Total differences: {dcount}.")
+        self.logger.info(f"Total config keys: {kcount}. Total differences: {dcount}.")
 
     def parse_line(self, line: str) -> UBXMessage:
         """
@@ -174,9 +177,9 @@ class UBXCompare:
                             self.get_attrs(cfgdict, str(parsed), fileno)
                             i += 1
         except Exception as err:
-            print(f"\nERROR parsing {filename}! \n{err}")
+            self.logger.error(f"ERROR parsing {filename}! \n{err}")
 
-        print(f"\n{i} configuration commands processed in {filename}")
+        self.logger.info(f"\n{i} configuration commands processed in {filename}")
 
 
 def main():
@@ -214,8 +217,9 @@ def main():
         default=1,
     )
 
-    args = ap.parse_args()
-    UBXCompare(args.infiles, args.format, args.diffsonly)
+    kwargs = set_common_args("ubxcompare", ap, logdefault=VERBOSITY_HIGH)
+
+    UBXCompare(kwargs.pop("infiles"), kwargs.pop("format"), kwargs.pop("diffsonly"))
 
 
 if __name__ == "__main__":

--- a/src/pygnssutils/ubxload.py
+++ b/src/pygnssutils/ubxload.py
@@ -14,7 +14,7 @@ configuration of the source device.
 
 Usage (all kwargs are optional):
 
-> ubxload port=/dev/ttyACM1 baud=9600 timeout=0.05 infile=ubxconfig.ubx verbose=1
+> ubxload --port /dev/ttyACM1 --baud 9600 --timeout 0.05 --infile ubxconfig.ubx --verbosity 1
 
 Created on 06 Jan 2023
 

--- a/src/pygnssutils/ubxload.py
+++ b/src/pygnssutils/ubxload.py
@@ -30,8 +30,16 @@ from datetime import datetime, timedelta
 from math import ceil
 from queue import Queue
 from threading import Event, Lock, Thread
+from time import sleep
 
-from pyubx2 import SET, UBX_PROTOCOL, UBXMessageError, UBXParseError, UBXReader
+from pyubx2 import (
+    NMEA_PROTOCOL,
+    SETPOLL,
+    UBX_PROTOCOL,
+    UBXMessageError,
+    UBXParseError,
+    UBXReader,
+)
 from serial import Serial
 
 from pygnssutils._version import __version__ as VERSION
@@ -39,16 +47,13 @@ from pygnssutils.globals import EPILOG
 
 ACK = "ACK-ACK"
 NAK = "ACK-NAK"
-# try increasing these values if device response is too slow:
-DELAY = 0.05  # delay between sends
-WRAPUP = 5  # final wrap up delay
 WAITTIME = 5  # wait time for acknowledgements
 
 
 class UBXLoader:
     """UBX Configuration Loader Class."""
 
-    def __init__(self, file: object, stream: object, **kwargs):
+    def __init__(self, filename: str, stream: object, **kwargs):
         """
         Constructor.
 
@@ -56,77 +61,53 @@ class UBXLoader:
         :param object stream: output serial stream
         """
 
-        self._file = file
+        self._filename = filename
         self._stream = stream
         self._waittime = ceil(kwargs.get("waittime", WAITTIME))
         self._verbose = int(kwargs.get("verbosity", 1))
-        self._ubxreader = UBXReader(self._stream, protfilter=UBX_PROTOCOL)
-        self._ubxloader = UBXReader(self._file, protfilter=UBX_PROTOCOL, msgmode=SET)
-
+        self._ubxreader = UBXReader(
+            self._stream, protfilter=NMEA_PROTOCOL | UBX_PROTOCOL
+        )
         self._serial_lock = Lock()
         self._out_queue = Queue()
         self._stop_event = Event()
         self._last_ack = datetime.now()
-
-        self._write_thread = Thread(
-            target=self._write_data,
-            daemon=True,
-            args=(
-                stream,
-                self._out_queue,
-                self._serial_lock,
-            ),
-        )
         self._read_thread = Thread(
             target=self._read_data,
             daemon=True,
             args=(
                 stream,
                 self._ubxreader,
-                self._serial_lock,
+                self._out_queue,
                 self._stop_event,
             ),
         )
 
         self._msg_ack = self._msg_nak = self._msg_write = self._msg_load = 0
 
-    def _load_data(self, ubr: UBXReader, queue: Queue):
+    def _load_data(self, filename: str, queue: Queue):
         """
         Get CFG-VALSET data from file and place it on output queue.
         """
 
-        eof = False
-        while not eof:
-            (_, parsed_data) = ubr.read()
-            if parsed_data is None:
-                eof = True
-            else:
-                self._msg_load += 1
-                queue.put(parsed_data)
-                if self._verbose > 1:
-                    print(f"LOAD {self._msg_load} - {parsed_data.identity}")
-
-    def _write_data(self, stream: object, queue: Queue, lock: Lock):
-        """
-        Read output queue and send CFG-VALSET messages to device.
-        """
-
-        while True:
-            parsed_data = queue.get()
-            if parsed_data is not None:
-                lock.acquire()
-                stream.write(parsed_data.serialize())
-                lock.release()
-                self._msg_write += 1
-                if self._verbose > 1:
-                    print(f"WRITE {self._msg_write} - {parsed_data.identity}")
-            queue.task_done()
+        with open(filename, "rb") as stream:
+            ubl = UBXReader(stream, msgmode=SETPOLL)
+            eof = False
+            while not eof:
+                (raw_data, parsed_data) = ubl.read()
+                if raw_data is None:
+                    eof = True
+                else:
+                    self._msg_load += 1
+                    queue.put(parsed_data)
+                    if self._verbose > 1:
+                        print(f"LOAD {self._msg_load} - {parsed_data.identity}")
 
     def _read_data(
         self,
         stream: object,
         ubr: UBXReader,
-        lock: Lock,
+        queue: Queue,
         stop: Event,
     ):
         """
@@ -137,41 +118,48 @@ class UBXLoader:
         # read until expected no of acknowledgements has been received
         # or waittime has been exceeded.
         while not stop.is_set():
-            if stream.in_waiting:
-                try:
-                    lock.acquire()
-                    (_, parsed_data) = ubr.read()
-                    lock.release()
-                    if parsed_data is not None:
-                        if (
-                            parsed_data.identity in (ACK, NAK)
-                            and parsed_data.clsID == 6  # CFG
-                            and parsed_data.msgID == 138  # CFG-VALSET
-                        ):
-                            self._last_ack = datetime.now()
-                            if parsed_data.identity == ACK:
-                                self._msg_ack += 1
-                            else:
-                                self._msg_nak += 1
-                            if self._verbose > 1:
-                                print(
-                                    (
-                                        "ACKNOWLEDGEMENT "
-                                        f"{self._msg_ack + self._msg_nak} - {parsed_data}"
-                                    )
-                                )
+            try:
+                (_, parsed_data) = ubr.read()
+                if parsed_data is not None:
                     if (
-                        self._msg_ack + self._msg_nak >= self._msg_load
-                        or datetime.now()
-                        > self._last_ack + timedelta(seconds=self._waittime)
+                        parsed_data.identity in (ACK, NAK)
+                        and parsed_data.clsID == 6  # CFG
+                        and parsed_data.msgID == 138  # CFG-VALSET
                     ):
-                        stop.set()
-                except (UBXMessageError, UBXParseError):
-                    continue
-                except Exception as err:
-                    if not stop.is_set():
-                        print(f"\n\nSomething went wrong {err}\n\n")
-                    continue
+                        self._last_ack = datetime.now()
+                        if parsed_data.identity == ACK:
+                            self._msg_ack += 1
+                        else:
+                            self._msg_nak += 1
+                        if self._verbose > 1:
+                            print(
+                                "ACKNOWLEDGEMENT "
+                                f"{self._msg_ack + self._msg_nak} - {parsed_data}"
+                            )
+
+                # send config message(s) to receiver
+                if not queue.empty():
+                    while not queue.empty():
+                        parsed_data = queue.get()
+                        self._msg_write += 1
+                        if self._verbose > 1:
+                            print(f"WRITE {self._msg_write} {parsed_data.identity}")
+                        stream.write(parsed_data.serialize())
+                    queue.task_done()
+
+                if (
+                    self._msg_ack + self._msg_nak >= self._msg_load
+                    or datetime.now()
+                    > self._last_ack + timedelta(seconds=self._waittime)
+                ):
+                    stop.set()
+
+            except (UBXMessageError, UBXParseError):
+                continue
+            except Exception as err:
+                if not stop.is_set():
+                    print(f"\n\nSomething went wrong {err}\n\n")
+                continue
 
     def run(self):
         """
@@ -181,25 +169,24 @@ class UBXLoader:
         rc = 1
         if self._verbose:
             print(
-                f"\nLoading configuration from {self._file.name} to {self._stream.port} ...",
+                f"\nLoading configuration from {self._filename} to {self._stream.port} ...",
                 "\nPress Ctrl-C to terminate early.",
             )
 
-        self._load_data(self._ubxloader, self._out_queue)
-        self._write_thread.start()
+        self._load_data(self._filename, self._out_queue)
         self._read_thread.start()
 
         # loop until all commands sent or user presses Ctrl-C
         while not self._stop_event.is_set():
             try:
-                pass
+                sleep(1)
             except KeyboardInterrupt:  # capture Ctrl-C
                 print(
                     "\n\nTerminated by user. WARNING! Configuration may be incomplete."
                 )
                 self._stop_event.set()
 
-        self._out_queue.join()
+        self._read_thread.join()
 
         if self._msg_ack == self._msg_load:
             if self._verbose:
@@ -256,7 +243,7 @@ def main():
         required=False,
         help="Wait time in seconds",
         type=float,
-        default=5,
+        default=WAITTIME,
     )
     ap.add_argument(
         "--verbosity",
@@ -269,12 +256,11 @@ def main():
 
     args = ap.parse_args()
 
-    with open(args.infile, "rb") as infile:
-        with Serial(args.port, args.baudrate, timeout=args.timeout) as serial_stream:
-            ubl = UBXLoader(
-                infile, serial_stream, verbosity=args.verbosity, waittime=args.waittime
-            )
-            ubl.run()
+    with Serial(args.port, args.baudrate, timeout=args.timeout) as serial_stream:
+        ubl = UBXLoader(
+            args.infile, serial_stream, verbosity=args.verbosity, waittime=args.waittime
+        )
+        ubl.run()
 
 
 if __name__ == "__main__":

--- a/src/pygnssutils/ubxload.py
+++ b/src/pygnssutils/ubxload.py
@@ -12,10 +12,6 @@ The binary file is created using the ubxsave utility and contains
 a series of CFG-VALSET messages representing the complete
 configuration of the source device.
 
-Usage (all kwargs are optional):
-
-> ubxload --port /dev/ttyACM1 --baud 9600 --timeout 0.05 --infile ubxconfig.ubx --verbosity 1
-
 Created on 06 Jan 2023
 
 :author: semuadmin
@@ -27,6 +23,7 @@ Created on 06 Jan 2023
 
 from argparse import ArgumentDefaultsHelpFormatter, ArgumentParser
 from datetime import datetime, timedelta
+from logging import getLogger
 from math import ceil
 from queue import Queue
 from threading import Event, Lock, Thread
@@ -43,17 +40,19 @@ from pyubx2 import (
 from serial import Serial
 
 from pygnssutils._version import __version__ as VERSION
-from pygnssutils.globals import EPILOG
+from pygnssutils.globals import EPILOG, VERBOSITY_HIGH
+from pygnssutils.helpers import progbar, set_common_args
 
 ACK = "ACK-ACK"
 NAK = "ACK-NAK"
 WAITTIME = 5  # wait time for acknowledgements
+WRITETIME = 0.02  # wait time between writes
 
 
 class UBXLoader:
     """UBX Configuration Loader Class."""
 
-    def __init__(self, filename: str, stream: object, **kwargs):
+    def __init__(self, stream: object, **kwargs):
         """
         Constructor.
 
@@ -61,10 +60,10 @@ class UBXLoader:
         :param object stream: output serial stream
         """
 
-        self._filename = filename
+        self.logger = getLogger(__name__)
         self._stream = stream
+        self._filename = kwargs["infile"]
         self._waittime = ceil(kwargs.get("waittime", WAITTIME))
-        self._verbose = int(kwargs.get("verbosity", 1))
         self._ubxreader = UBXReader(
             self._stream, protfilter=NMEA_PROTOCOL | UBX_PROTOCOL
         )
@@ -100,8 +99,7 @@ class UBXLoader:
                 else:
                     self._msg_load += 1
                     queue.put(parsed_data)
-                    if self._verbose > 1:
-                        print(f"LOAD {self._msg_load} - {parsed_data.identity}")
+                    self.logger.debug(f"LOAD {self._msg_load} - {parsed_data.identity}")
 
     def _read_data(
         self,
@@ -131,20 +129,24 @@ class UBXLoader:
                             self._msg_ack += 1
                         else:
                             self._msg_nak += 1
-                        if self._verbose > 1:
-                            print(
-                                "ACKNOWLEDGEMENT "
-                                f"{self._msg_ack + self._msg_nak} - {parsed_data}"
-                            )
+                        self.logger.debug(
+                            f"ACKNOWLEDGEMENT {self._msg_ack + self._msg_nak} - {parsed_data}"
+                        )
 
                 # send config message(s) to receiver
                 if not queue.empty():
+                    i = 0
                     while not queue.empty():
+                        i += 1
+                        n = ceil(50 / self._msg_load)
+                        progbar(i * n, self._msg_load * n, 50)
                         parsed_data = queue.get()
                         self._msg_write += 1
-                        if self._verbose > 1:
-                            print(f"WRITE {self._msg_write} {parsed_data.identity}")
+                        self.logger.debug(
+                            f"WRITE {self._msg_write} {parsed_data.identity}"
+                        )
                         stream.write(parsed_data.serialize())
+                        sleep(WRITETIME)
                     queue.task_done()
 
                 if (
@@ -158,7 +160,7 @@ class UBXLoader:
                 continue
             except Exception as err:
                 if not stop.is_set():
-                    print(f"\n\nSomething went wrong {err}\n\n")
+                    self.logger.error(f"Something went wrong {err}")
                 continue
 
     def run(self):
@@ -167,11 +169,10 @@ class UBXLoader:
         """
 
         rc = 1
-        if self._verbose:
-            print(
-                f"\nLoading configuration from {self._filename} to {self._stream.port} ...",
-                "\nPress Ctrl-C to terminate early.",
-            )
+        self.logger.info(
+            f"Loading configuration from {self._filename} to {self._stream.port}. "
+            "Press Ctrl-C to terminate early.",
+        )
 
         self._load_data(self._filename, self._out_queue)
         self._read_thread.start()
@@ -181,33 +182,35 @@ class UBXLoader:
             try:
                 sleep(1)
             except KeyboardInterrupt:  # capture Ctrl-C
-                print(
-                    "\n\nTerminated by user. WARNING! Configuration may be incomplete."
+                self.logger.warning(
+                    "Terminated by user. Configuration may be incomplete."
                 )
                 self._stop_event.set()
 
         self._read_thread.join()
 
         if self._msg_ack == self._msg_load:
-            if self._verbose:
-                print(
-                    "\nConfiguration successfully loaded.",
-                    f"\n{self._msg_load} CFG-VALSET messages sent and acknowledged.",
-                )
+            self.logger.info(
+                "Configuration successfully loaded. "
+                f"{self._msg_load} CFG-VALSET messages sent and acknowledged."
+            )
         else:
             null = self._msg_load - self._msg_ack - self._msg_nak
             rc = 0
-            if self._verbose:
-                print(
-                    "\nWARNING! Configuration was not successfully loaded.",
-                    f"\n{self._msg_load} CFG-VALSET messages sent,",
-                    f"{self._msg_ack} acknowledged, {self._msg_nak} rejected,",
-                    f"{null} null responses.",
+            self.logger.warning(
+                "Configuration may be incomplete. "
+                f"{self._msg_load} CFG-VALSET messages sent, "
+                f"{self._msg_ack} acknowledged, {self._msg_nak} rejected, "
+                f"{null} null responses."
+            )
+            if null:
+                self.logger.warning(
+                    f"Consider increasing waittime to >{self._waittime} seconds."
                 )
-                if null:
-                    print(f"Consider increasing waittime to >{self._waittime}.")
-                if self._msg_nak:
-                    print("Check device is compatible with this saved configuration.")
+            if self._msg_nak:
+                self.logger.warning(
+                    "Check device is compatible with this saved configuration."
+                )
 
         return rc
 
@@ -245,21 +248,13 @@ def main():
         type=float,
         default=WAITTIME,
     )
-    ap.add_argument(
-        "--verbosity",
-        required=False,
-        help="Verbosity 0 = low, 1 = medium, 2 = high, 3 = debug",
-        type=int,
-        choices=[0, 1, 2, 3],
-        default=1,
-    )
 
-    args = ap.parse_args()
+    kwargs = set_common_args("ubxload", ap, logdefault=VERBOSITY_HIGH)
 
-    with Serial(args.port, args.baudrate, timeout=args.timeout) as serial_stream:
-        ubl = UBXLoader(
-            args.infile, serial_stream, verbosity=args.verbosity, waittime=args.waittime
-        )
+    with Serial(
+        kwargs.pop("port"), kwargs.pop("baudrate"), timeout=kwargs.pop("timeout")
+    ) as serial_stream:
+        ubl = UBXLoader(serial_stream, **kwargs)
         ubl.run()
 
 

--- a/src/pygnssutils/ubxsave.py
+++ b/src/pygnssutils/ubxsave.py
@@ -19,7 +19,7 @@ waittime and/or baudrate or temporarily reducing periodic message rates.
 
 Usage (all kwargs are optional):
 
-> ubxsave port=/dev/ttyACM1 baud=9600 timeout=0.02 outfile=ubxconfig.ubx verbose=1
+> ubxsave --port /dev/ttyACM1 --baud 9600 --timeout 0.02 --outfile ubxconfig.ubx --verbosity 1
 
 Created on 06 Jan 2023
 

--- a/src/pygnssutils/ubxsave.py
+++ b/src/pygnssutils/ubxsave.py
@@ -31,6 +31,7 @@ Created on 06 Jan 2023
 # pylint: disable=invalid-name
 
 from argparse import ArgumentDefaultsHelpFormatter, ArgumentParser
+from logging import getLogger
 from math import ceil
 from queue import Queue
 from threading import Event, Lock, Thread
@@ -50,8 +51,8 @@ from pyubx2 import (
 from serial import Serial
 
 from pygnssutils._version import __version__ as VERSION
-from pygnssutils.globals import EPILOG
-from pygnssutils.helpers import progbar
+from pygnssutils.globals import EPILOG, VERBOSITY_HIGH
+from pygnssutils.helpers import progbar, set_common_args
 
 # try increasing these values if device response is too slow:
 DELAY = 0.02  # delay between polls
@@ -64,9 +65,9 @@ class UBXSaver:
     def __init__(self, file: object, stream: object, **kwargs):
         """Constructor."""
 
+        self.logger = getLogger(__name__)
         self._file = file
         self._stream = stream
-        self._verbose = int(kwargs.get("verbosity", 1))
         self._waittime = ceil(kwargs.get("waittime", WRAPUP))
 
         self._ubxreader = UBXReader(stream, protfilter=UBX_PROTOCOL)
@@ -120,8 +121,7 @@ class UBXSaver:
             lock.acquire()
             stream.write(message.serialize())
             self._msg_write += 1
-            if self._verbose > 1:
-                print(f"WRITE {self._msg_write} - {message.identity}")
+            self.logger.debug(f"WRITE {self._msg_write} - {message.identity}")
             lock.release()
             queue.task_done()
 
@@ -150,13 +150,12 @@ class UBXSaver:
                         if parsed_data.identity == "CFG-VALGET":
                             queue.put((raw_data, parsed_data))
                             self._msg_rcvd += 1
-                            if self._verbose > 1:
-                                print(
-                                    f"RESPONSE {self._msg_rcvd} - {parsed_data.identity}"
-                                )
+                            self.logger.debug(
+                                f"RESPONSE {self._msg_rcvd} - {parsed_data.identity}"
+                            )
             except Exception as err:
                 if not stop.is_set():
-                    print(f"\n\nSomething went wrong {err}\n\n")
+                    self.logger.error(f"Something went wrong {err}")
                 continue
 
     def _save_data(self, file: object, queue: Queue):
@@ -195,8 +194,7 @@ class UBXSaver:
         data = UBXMessage.config_set(
             layers=SET_LAYER_RAM, transaction=txn, cfgData=cfgdata
         )
-        if self._verbose > 1:
-            print(f"SAVE {self._msg_save} - {data.identity}")
+        self.logger.debug(f"SAVE {self._msg_save} - {data.identity}")
         file.write(data.serialize())
 
     def run(self):
@@ -205,11 +203,10 @@ class UBXSaver:
         """
 
         rc = 1
-        if self._verbose:
-            print(
-                f"\nSaving configuration from {self._stream.port} to {self._file.name} ..."
-            )
-            print("Press Ctrl-C to terminate early.")
+        self.logger.info(
+            f"Saving configuration from {self._stream.port} to {self._file.name}. "
+            "Press Ctrl-C to terminate early."
+        )
 
         # loop until all commands sent or user presses Ctrl-C
         try:
@@ -220,25 +217,22 @@ class UBXSaver:
             position = 0
             keys = []
             for i, key in enumerate(UBX_CONFIG_DATABASE):
-                if self._verbose == 1:
-                    progbar(i, len(UBX_CONFIG_DATABASE), 50)
+                progbar(i, len(UBX_CONFIG_DATABASE), 50)
                 keys.append(key)
                 msg = UBXMessage.config_poll(layer, position, keys)
                 self._send_queue.put(msg)
                 self._msg_sent += 1
-                if self._verbose > 1:
-                    print(f"POLL {i} - {msg.identity}")
+                self.logger.debug(f"POLL {i} - {msg.identity}")
                 keys = []
                 sleep(DELAY)
 
-            if self._verbose:
-                for i in range(self._waittime):
-                    print(
-                        f"Waiting {self._waittime - i} seconds for final responses..."
-                        + " " * 20,
-                        end="\r",
-                    )
-                    sleep(1)
+            for i in range(self._waittime):
+                print(
+                    f"Waiting {self._waittime - i} seconds for final responses..."
+                    + " " * 20,
+                    end="\r",
+                )
+                sleep(1)
             # sleep(self._waittime)
 
             self._stop_event.set()
@@ -248,25 +242,23 @@ class UBXSaver:
 
         except KeyboardInterrupt:  # capture Ctrl-C
             self._stop_event.set()
-            print("\n\nTerminated by user. WARNING! Configuration may be incomplete.")
+            self.logger.warning("Terminated by user. Configuration may be incomplete.")
 
         if self._msg_rcvd == self._cfgkeys:
-            if self._verbose:
-                print(
-                    "Configuration successfully saved." + " " * 15,
-                    f"\n{self._msg_save} CFG-VALSET messages saved to {self._file.name}",
-                )
+            self.logger.info(
+                "Configuration successfully saved. "
+                f"{self._msg_save} CFG-VALSET messages saved to {self._file.name}"
+            )
         else:
             rc = 0
-            if self._verbose:
-                print(
-                    "WARNING! Configuration not successfully saved",
-                    f"\n{self._msg_sent} CFG-VALGET polls sent to {self._stream.port}",
-                    f"\n{self._msg_rcvd} CFG-VALGET responses received",
-                    f"\n{self._msg_save} CFG-VALSET messages containing {self._cfgkeys} keys",
-                    f"({self._cfgkeys*100/self._msg_rcvd:.1f}%) written to {self._file.name}",
-                    f"\nConsider increasing waittime to >{self._waittime}.",
-                )
+            self.logger.warning(
+                "Configuration may not be successfully saved. "
+                f"{self._msg_sent} CFG-VALGET polls sent to {self._stream.port}, "
+                f"{self._msg_rcvd} CFG-VALGET responses received, "
+                f"{self._msg_save} CFG-VALSET messages containing {self._cfgkeys} keys "
+                f"({self._cfgkeys*100/self._msg_rcvd:.1f}%) written to {self._file.name}. "
+                f"Consider increasing waittime to >{self._waittime}."
+            )
 
         return rc
 
@@ -310,22 +302,14 @@ def main():
         type=int,
         default=WRAPUP,
     )
-    ap.add_argument(
-        "--verbosity",
-        required=False,
-        help="Verbosity 0 = low, 1 = medium, 2 = high, 3 = debug",
-        type=int,
-        choices=[0, 1, 2, 3],
-        default=1,
-    )
 
-    args = ap.parse_args()
+    kwargs = set_common_args("ubxsave", ap, logdefault=VERBOSITY_HIGH)
 
-    with open(args.outfile, "wb") as outfile:
-        with Serial(args.port, args.baudrate, timeout=args.timeout) as serial_stream:
-            ubs = UBXSaver(
-                outfile, serial_stream, verbosity=args.verbosity, waittime=args.waittime
-            )
+    with open(kwargs.pop("outfile"), "wb") as outfile:
+        with Serial(
+            kwargs.pop("port"), kwargs.pop("baudrate"), timeout=kwargs.pop("timeout")
+        ) as serial_stream:
+            ubs = UBXSaver(outfile, serial_stream, **kwargs)
             ubs.run()
 
 

--- a/src/pygnssutils/ubxsetrate.py
+++ b/src/pygnssutils/ubxsetrate.py
@@ -7,7 +7,8 @@ to a local serial port via the UBX CFG-MSG command.
 
 Usage:
 
-ubxsetrate --port dev/tty.usbmodem1301 --baudrate 38400 --timeout 5 --msgClass 0xf1 --msgID 0x00 --rate 1 --verbosity 1
+ubxsetrate --port dev/tty.usbmodem1301 --baudrate 38400 --timeout 5 
+   --msgClass 0xf1 --msgID 0x00 --rate 1 --verbosity 1
 
 or (to turn off all NMEA messages):
 
@@ -23,6 +24,7 @@ Created on 12 Dec 2022
 # pylint: disable=invalid-name
 
 from argparse import ArgumentDefaultsHelpFormatter, ArgumentParser
+from logging import getLogger
 
 from pyubx2 import SET, UBX_CLASSES, UBX_MSGIDS, UBXMessage
 from serial import Serial
@@ -39,7 +41,9 @@ from pygnssutils.globals import (
     MINNMEA,
     MINUBX,
     MINUBX_ID,
+    VERBOSITY_HIGH,
 )
+from pygnssutils.helpers import set_common_args
 
 
 class UBXSetRate:
@@ -66,6 +70,7 @@ class UBXSetRate:
         :raises: ParameterError
         """
 
+        self.logger = getLogger(__name__)
         try:
             self._serialOut = None
             self._port = kwargs.get("port")
@@ -105,7 +110,9 @@ class UBXSetRate:
         """
 
         try:
-            print(f"Opening serial port {self._port} @ {self._baudrate} baud ...\n")
+            self.logger.info(
+                f"Opening serial port {self._port} @ {self._baudrate} baud ..."
+            )
             self._serialOut = Serial(self._port, self._baudrate, timeout=self._timeout)
 
             if self._msgClass == ALLNMEA:  # all available NMEA messages
@@ -135,7 +142,7 @@ class UBXSetRate:
             raise err from err
         finally:
             if self._serialOut is not None:
-                print("Configuration message(s) sent.")
+                self.logger.info("Configuration message(s) sent.")
                 self._serialOut.close()
 
     def _sendmsg(self, msgClass: int, msgID: int):
@@ -159,7 +166,7 @@ class UBXSetRate:
             rateSPI=self._rate,
         )
 
-        print(f"Sending configuration message {msg}...\n")
+        self.logger.info(f"Sending configuration message {msg}...")
         self._serialOut.write(msg.serialize())
 
 
@@ -209,7 +216,7 @@ def main():
         default=1,
     )
 
-    kwargs = vars(ap.parse_args())
+    kwargs = set_common_args("ubxsetrate", ap, logdefault=VERBOSITY_HIGH)
 
     try:
         usr = UBXSetRate(**kwargs)

--- a/src/pygnssutils/ubxsetrate.py
+++ b/src/pygnssutils/ubxsetrate.py
@@ -7,11 +7,11 @@ to a local serial port via the UBX CFG-MSG command.
 
 Usage:
 
-ubxsetrate port=/dev/tty.usbmodem1301 baudrate=38400 timeout=5 msgClass=0xf1 msgID=0x00 rate=1
+ubxsetrate --port dev/tty.usbmodem1301 --baudrate 38400 --timeout 5 --msgClass 0xf1 --msgID 0x00 --rate 1 --verbosity 1
 
-or:
+or (to turn off all NMEA messages):
 
-ubxsetrate port=/dev/tty.usbmodem1301 msgClass=allnmea
+ubxsetrate --port /dev/tty.usbmodem1301 --msgClass allnmea --rate 0
 
 Created on 12 Dec 2022
 
@@ -74,7 +74,7 @@ class UBXSetRate:
             mcls = kwargs.get("msgClass")
             mid = kwargs.get("msgID", 0)
 
-            if mcls in (ALLNMEA, ALLUBX, MINNMEA, MINUBX):
+            if mcls.lower() in (ALLNMEA, ALLUBX, MINNMEA, MINUBX):
                 self._msgClass = mcls
             else:
                 self._msgClass = int(mcls, 16) if mcls[0:2] == "0x" else int(mcls)

